pkg/clients/azure/aks/aks.go

```diff
@@ -0,0 +1,146 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aks
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/services/containerservice/mgmt/2018-03-31/containerservice"
+	"github.com/Azure/go-autorest/autorest/to"
+	computev1alpha1 "github.com/crossplaneio/crossplane/pkg/apis/azure/compute/v1alpha1"
+	"github.com/crossplaneio/crossplane/pkg/clients/azure"
+	"github.com/crossplaneio/crossplane/pkg/util"
+)
+
+const (
+	AgentPoolProfileNameFmt = "%s-nodepool"
+	MaxClusterNameLength    = 31
+)
+
+// Client interface for AKS Cluster client
+type Client interface {
+	Create(string, computev1alpha1.AKSClusterSpec) (containerservice.ManagedCluster, error)
+	Get(string, string) (containerservice.ManagedCluster, error)
+	Delete(string, string) error
+	ListCredentials(string, string) (containerservice.CredentialResult, error)
+}
+
+// NewAKSClient return AKS client implementation
+func NewAKSClient(config *azure.ClientCredentialsConfig) (Client, error) {
+	authorizer, err := config.Authorizer()
+	if err != nil {
+		return nil, err
+	}
+
+	client := containerservice.NewManagedClustersClient(config.SubscriptionID)
+	client.Authorizer = authorizer
+
+	if err := client.AddToUserAgent(azure.UserAgent); err != nil {
+		return nil, err
+	}
+
+	return &AKSClient{
+		client:                  &client,
+		ClientCredentialsConfig: config,
+	}, nil
+}
+
+// AKSClient implementing Client interface
+type AKSClient struct {
+	client *containerservice.ManagedClustersClient
+	*azure.ClientCredentialsConfig
+}
+
+// Create new AKS Cluster with given name and specs
+// **NOTE**: the actual cluster name can be different from the provided name
+func (a *AKSClient) Create(name string, spec computev1alpha1.AKSClusterSpec) (containerservice.ManagedCluster, error) {
+	name = trimName(name)
+
+	nodeCount := int32(computev1alpha1.DefaultNodeCount)
+	if spec.NodeCount != nil {
+		nodeCount = int32(*spec.NodeCount)
+	}
+
+	agentPoolProfileName := fmt.Sprintf(AgentPoolProfileNameFmt, name)
+	enableRBAC := !spec.DisableRBAC
+
+	// create (or update) - for now we care about creation only
+	_, err := a.client.CreateOrUpdate(
+		context.Background(),
+		spec.ResourceGroupName,
+		name,
+		containerservice.ManagedCluster{
+			Name:     &name,
+			Location: util.String(spec.Location),
+			ManagedClusterProperties: &containerservice.ManagedClusterProperties{
+				KubernetesVersion: &spec.Version,
+				DNSPrefix:         &name,
+				AgentPoolProfiles: &[]containerservice.ManagedClusterAgentPoolProfile{
+					{
+						Count:  to.Int32Ptr(nodeCount),
+						Name:   to.StringPtr(agentPoolProfileName),
+						VMSize: containerservice.VMSizeTypes(spec.NodeVMSize),
+					},
+				},
+				ServicePrincipalProfile: &containerservice.ManagedClusterServicePrincipalProfile{
```

Comment:
this is giving the AKS cluster the Azure administrator credentials with full access to the entirety of their Azure account.  All the documentation I found when implementing this originally said to create a new and specific AD app and service principal for the AKS cluster to use. For example:
https://docs.microsoft.com/en-us/azure/aks/tutorial-kubernetes-deploy-cluster#create-a-service-principal

The Azure CLI and the Azure Portal UI also creates a new app/service principal for the cluster to use as well.  Do you think it's OK to give the AKS cluster full admin creds to the whole Azure account?

Reply:
@jbw976 maybe I do not entirely follow, can you elaborate what do you mean by:
> this is giving the AKS cluster the Azure administrator credentials with full access to the entirety of their Azure account.

Specifically, what is not clear is there is nothing actually "given" to the AKS cluster. The cluster is created with the admin user account, but it does not receive or store admin user credentials as far as I can tell. There must be something I am missing. Is there an implication that there could be a security vulnerability? AFAIK, creating the AKS cluster w/ the admin account creds is (or should be) no less secure as creating say AD app (with the same admin account creds).

With that said. I do see a potential value in providing a dedicated Resource Group and AD Application for a set of managed resources. However, I think this should be handled separately, either in offline mode (similar how we handled Resource Group creation) or better, with added support for the Resource Group and AD Apps as the `Crossplane` types. In either approach, the main point here - the Resource Group or AD Apps creation shouldn't be "bundled" or be on the same execution path as AKS cluster creation. 

As for the example, I modeled this based on [ExampleCreateAKS in azure-sdk-for-go-samples](https://github.com/Azure-Samples/azure-sdk-for-go-samples/blob/master/compute/container_cluster_test.go)